### PR TITLE
Add metrics support for poller and scaler events

### PR DESCRIPTION
**Pull request by nielm**
Add the following metrics to google-cloud-monitoring: 

----

Scaler:
1) cloudspannerecosystem/autoscaler/scaler/scaling-success
2) cloudspannerecosystem/autoscaler/scaler/scaling-denied
3) cloudspannerecosystem/autoscaler/scaler/scaling-failed
4) cloudspannerecosystem/autoscaler/scaler/requests-success
5) cloudspannerecosystem/autoscaler/scaler/requests-failed

first 3 have attributes: 
* spanner_project_id
* spanner_instance_id
* scaling_method
* scaling_direction: SCALE_UP, SCALE_DOWN, SCALE_SAME

scaling-denied additionally has a scaling_denied_reason attribute: 
* MAX_SIZE
* SAME_SIZE
* WITHIN_COOLDOWN

----

Poller

1) cloudspannerecosystem/autoscaler/poller/polling-success
3) cloudspannerecosystem/autoscaler/poller/polling-failed
4) cloudspannerecosystem/autoscaler/poller/requests-success
5) cloudspannerecosystem/autoscaler/poller/requests-failed

first 2 have attributes: 
* spanner_project_id
* spanner_instance_id

### CODE REVIEW

File: terraform/gke/README.md [hunk 1]
```diff
@@ -93,6 +94,11 @@ infrastructure and configuration of their own Autoscalers on Kubernetes.
 8.  If the configured cooldown period has passed, then the Scaler requests the
     Spanner Instance to scale out or in.
 
+9.  Both Poller and Scaler publish counters to an [OpenTelemetry Collector][otel-collector],
+    also running in Kubernetes, which is configure to forward these counters to
```
Comment (henrybell) [inline terraform/gke/README.md:98]: Nit: "configured"

File: src/autoscaler-common/counters_base.js [hunk 2]
```diff
@@ -0,0 +1,448 @@
...
+      } else {
+        logger.warn('WARNING: running under Kubernetes, but K8S_POD_NAME ' +
+        'environment variable is not set. ' +
+        'This may lead to Send TimeSeries errors');
+      }
+    }
+
+    const gcpResources = new GcpDetectorSync().detect();
+    await gcpResources.waitForAsyncAttributes();
+    logger.debug('Got gcpResources attrs:1 %o', gcpResources.attributes);
```
Comment (henrybell) [inline src/autoscaler-common/counters_base.js:231]: Nit: the formatting of the object in the debug message body takes up multiline real-estate in the top-level logging views in the web console -- do you think it's worth embedding this in a different field?

File: src/autoscaler-common/counters_base.js [hunk 3]
```diff
@@ -0,0 +1,448 @@
...
+        logger.warn('WARNING: running under Cloud Functions, but FAAS_ID ' +
+        'resource attribute is not set. ' +
+        'This may lead to Send TimeSeries errors');
+      }
+    }
+
+    const resources = gcpResources.merge(new Resource(RESOURCE_ATTRIBUTES));
+    await resources.waitForAsyncAttributes();
+
+    logger.debug('Got metrics resource attrs2: %o', resources.attributes);
```
Comment (henrybell) [inline src/autoscaler-common/counters_base.js:254]: Nit: the formatting of the object in the debug message body takes up multiline real-estate in the top-level logging views in the web console -- do you think it's worth embedding this in a different field?

**Review by henrybell — CHANGES_REQUESTED** [review]

File: README.md [hunk 4]
```diff
@@ -100,6 +100,62 @@ In both of the above instances, the Google Cloud Platform resources are
...
+        failed polls of the Spanner instance metrics.
+    *   Both of these metrics have `projectid` and `instanceid` to identify the
+        Spanner instance.
+
+### Scaler
+
+*   Message processing counters:
+    *   `cloudspannerecosystem/autoscaler/scaler/requests-success` - the number
+        of scaling request messages recieved and processed successfully.
+    *   `cloudspannerecosystem/autoscaler/scaler/requests-success` - the number
```
[the remark this answers is not in the document]
Comment (nielm) [reply README.md:129]: Done

File: README.md [hunk 5]
```diff
@@ -100,6 +100,62 @@ In both of the above instances, the Google Cloud Platform resources are
...
+
+The autoscaler publishes the following metrics to Cloud Monitoring which can be
+used to monitor the behavior of the autoscaler, and to configure alerts.
+
+### Poller
+
+*   Message processing counters:
+    *   `cloudspannerecosystem/autoscaler/poller/requests-success` - the number
+        of polling request messages recieved and processed successfully.
+    *   `cloudspannerecosystem/autoscaler/poller/requests-success` - the number
```
[the remark this answers is not in the document]
Comment (nielm) [reply README.md:113]: Done

File: src/autoscaler-common/counters_base.js (hunk 2 above)
Comment (nielm) [reply src/autoscaler-common/counters_base.js:231]: Also, that was debug code I left in! Removing!

File: src/autoscaler-common/counters_base.js (hunk 3 above)
Comment (nielm) [reply src/autoscaler-common/counters_base.js:254]: removed

File: terraform/gke/README.md (hunk 1 above)
Comment (nielm) [reply terraform/gke/README.md:98]: Done

File: terraform/gke/README.md [hunk 6]
```diff
@@ -608,6 +608,18 @@ A [NetworkPolicy rule](../../kubernetes/decoupled/autoscaler-pkg/networkpolicy.y
 is also configured to allow traffic from the `poller` and `scaler` workloads
 (labelled with `otel-submitter:true`) to the `otel-collector` service.
 
+If the environmental variable `OTLP_COLLECTOR_URL` is not specified, the metrics
```
Comment (nielm) [inline terraform/gke/README.md:611]: @henrybell Added this since your last review.

Comment (henrybell) [reply terraform/gke/README.md:611]: Final nit: "environmental" -> "environment"!

Comment (nielm) [reply terraform/gke/README.md:611]: Done

**Pull request merged**